{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CrQzzQv8ju-P"
      },
      "source": [
        "In this exercise, we want to familiarize with the library biopython and working with sequences.\n",
        "For this, we will download a sequence file and work with it.\n",
        "You can choose any sequence you like (DNA) or use the one we suggest further down.\n",
        "\n",
        "**Working with Genbank files**\n",
        "\n",
        "We work with files from the genebank.\n",
        "You can find more information about what the genebank is in the below link:\n",
        "\n",
        "https://en.wikipedia.org/wiki/GenBank"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9PIpe1tY5jrV"
      },
      "source": [
        "First we need to install the prerequisites (biopython)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "_lUNSgf5hh3W"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.85-cp312-cp312-win_amd64.whl.metadata (13 kB)\n",
            "Collecting numpy (from biopython)\n",
            "  Downloading numpy-2.3.4-cp312-cp312-win_amd64.whl.metadata (60 kB)\n",
            "     ---------------------------------------- 0.0/60.9 kB ? eta -:--:--\n",
            "     ------ --------------------------------- 10.2/60.9 kB ? eta -:--:--\n",
            "     ------------------------------- ------ 51.2/60.9 kB 871.5 kB/s eta 0:00:01\n",
            "     ------------------------------- ------ 51.2/60.9 kB 871.5 kB/s eta 0:00:01\n",
            "     -------------------------------------- 60.9/60.9 kB 465.1 kB/s eta 0:00:00\n",
            "Downloading biopython-1.85-cp312-cp312-win_amd64.whl (2.8 MB)\n",
            "   ---------------------------------------- 0.0/2.8 MB ? eta -:--:--\n",
            "   --- ------------------------------------ 0.3/2.8 MB 8.3 MB/s eta 0:00:01\n",
            "   ---------- ----------------------------- 0.7/2.8 MB 9.1 MB/s eta 0:00:01\n",
            "   -------------------------- ------------- 1.9/2.8 MB 17.3 MB/s eta 0:00:01\n",
            "   ---------------------------------------  2.8/2.8 MB 20.0 MB/s eta 0:00:01\n",
            "   ---------------------------------------- 2.8/2.8 MB 16.3 MB/s eta 0:00:00\n",
            "Downloading numpy-2.3.4-cp312-cp312-win_amd64.whl (12.8 MB)\n",
            "   ---------------------------------------- 0.0/12.8 MB ? eta -:--:--\n",
            "   ------ --------------------------------- 2.0/12.8 MB 62.8 MB/s eta 0:00:01\n",
            "   ------------ --------------------------- 3.9/12.8 MB 49.5 MB/s eta 0:00:01\n",
            "   ---------------- ----------------------- 5.4/12.8 MB 42.8 MB/s eta 0:00:01\n",
            "   ---------------------- ----------------- 7.1/12.8 MB 38.2 MB/s eta 0:00:01\n",
            "   ---------------------------- ----------- 9.2/12.8 MB 39.3 MB/s eta 0:00:01\n",
            "   ---------------------------------------  12.8/12.8 MB 43.7 MB/s eta 0:00:01\n",
            "   ---------------------------------------  12.8/12.8 MB 43.7 MB/s eta 0:00:01\n",
            "   ---------------------------------------  12.8/12.8 MB 43.7 MB/s eta 0:00:01\n",
            "   ---------------------------------------- 12.8/12.8 MB 31.2 MB/s eta 0:00:00\n",
            "Installing collected packages: numpy, biopython\n",
            "Successfully installed biopython-1.85 numpy-2.3.4\n",
            "Note: you may need to restart the kernel to use updated packages.\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "\n",
            "[notice] A new release of pip is available: 24.0 -> 25.2\n",
            "[notice] To update, run: python.exe -m pip install --upgrade pip\n"
          ]
        }
      ],
      "source": [
        "%pip install biopython"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "h_G0a6S6WgP9"
      },
      "source": [
        "1. Write python codes using the biopython package that reads a genbank file. (2 Points)\n",
        "\n",
        "You can check this https://www.ncbi.nlm.nih.gov/genbank/ website and choose whatever sequence you want and store the sequence file in any format you want. Please choose a sequence with less than 1000 basepairs or we will have problems with moodle upload limits. You can also just use the following sequence:\n",
        "https://www.ncbi.nlm.nih.gov/nuccore/AB026718.1\n",
        "\n",
        "Hint: read the file into a SeqRecord Object"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "id": "bGDTB5ZJWpK5"
      },
      "outputs": [],
      "source": [
        "from Bio import SeqIO\n",
        "record = SeqIO.read(\"sequence.gb\", \"genbank\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VD6tokPP5xgV"
      },
      "source": [
        "2. Print the identifier, name and description of the sequence (2 Points)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8Re-gu2-O3eh"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ID: AB026718.1\n",
            "Name: AB026718\n",
            "Description: Malus domestica mitochondrial DNA, nad4L-orf25 intergenic sequence\n"
          ]
        }
      ],
      "source": [
        "\n",
        "print(\"ID:\", record.id)\n",
        "print(\"Name:\", record.name)\n",
        "print(\"Description:\", record.description)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0aVkD1Hc56Ve"
      },
      "source": [
        "3. Print the first 100 (or all if your sequence has less than 100) characters of the sequence (1 Point)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "8665pdBE56i6"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "First 100 bases of the sequence:\n",
            "GGTTAAACATGACTCCTAGAGAATTACAAAGAAGTTCTCTTCTCCTTTCGTTCTCTTCTTTCTTTTTTTGGTTGGCAGGGTCAGGGCCTTTCTCGCTGGG\n"
          ]
        }
      ],
      "source": [
        "sequence = record.seq\n",
        "\n",
        "length_to_show = min(100, len(sequence))\n",
        "\n",
        "print(\"First\", length_to_show, \"bases of the sequence:\")\n",
        "print(sequence[:length_to_show])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ULdihqS_56t7"
      },
      "source": [
        "4. Print the number of external references (dbxrefs) and ids of the external references. (1 Point)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "id": "brWv5xRu5627"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Number of external references: 0\n",
            "External reference IDs:\n"
          ]
        }
      ],
      "source": [
        "dbxrefs = record.dbxrefs\n",
        "\n",
        "print(\"Number of external references:\", len(dbxrefs))\n",
        "print(\"External reference IDs:\")\n",
        "for ref in dbxrefs:\n",
        "    print(ref)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BMkr3AuT56_2"
      },
      "source": [
        "5. Print the name of the organism (hint: check the annotations dictionary at the key “organism”) (1 Point)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "Hg9kmkIt57J6"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Organism name: Malus domestica\n"
          ]
        }
      ],
      "source": [
        "organism_name = record.annotations.get(\"organism\", \"Unknown organism\")\n",
        "\n",
        "print(\"Organism name:\", organism_name)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DDI4_j8w57Ry"
      },
      "source": [
        "6. Retrieve and print all (if any) associated publications (hint: annotation dictionary, key:”references”) (1 Point)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "PuSQiXZC57do"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Associated publications:\n",
            "- Title: The nad4L-orf25 gene cluster is conserved and expressed in sugar beet mitochondria\n",
            "  Authors: Kubo,T., Yamamoto,M.P. and Mikami,T.\n",
            "  Journal: Theor. Appl. Genet. 100, 214-220 (2000)\n",
            "\n",
            "- Title: Direct Submission\n",
            "  Authors: Kubo,T. and Mikami,T.\n",
            "  Journal: Submitted (26-APR-1999) Tomohiko Kubo, Faculty of Agriculture, Hokkaido University, Laboratory of Genetic Engineering; N-9 W-9, Kita-ku, Sapporo, Hokkaido 060-8589, Japan (E-mail:tomohiko@abs.agr.hokudai.ac.jp, Tel:81-11-706-2806, Fax:81-11-716-0879)\n",
            "\n"
          ]
        }
      ],
      "source": [
        "references = record.annotations.get(\"references\", [])\n",
        "\n",
        "if references:\n",
        "    print(\"Associated publications:\")\n",
        "    for ref in references:\n",
        "        print(f\"- Title: {ref.title}\")\n",
        "        print(f\"  Authors: {ref.authors}\")\n",
        "        print(f\"  Journal: {ref.journal}\")\n",
        "        print()\n",
        "else:\n",
        "    print(\"No associated publications found.\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "X3Hpm7_26oYu"
      },
      "source": [
        "7. Retrieve and print all the locations of “CDS” features of the sequence (hint: check the features) (1 Point)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "o_3lIHHI6onV"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Locations of CDS features:\n",
            "No CDS features found in this sequence.\n"
          ]
        }
      ],
      "source": [
        "print(\"Locations of CDS features:\")\n",
        "\n",
        "cds_features = [feature for feature in record.features if feature.type == \"CDS\"]\n",
        "\n",
        "if cds_features:\n",
        "    for i, feature in enumerate(cds_features, start=1):\n",
        "        print(f\"CDS {i}: {feature.location}\")\n",
        "else:\n",
        "    print(\"No CDS features found in this sequence.\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ldj0l1sSH4zt"
      },
      "source": [
        "8. Create a function (using Biopython or without it) that computes the reverse complement of a given DNA sequence. Ensure that the input sequence is validated to contain only valid nucleotide characters (A, T, C, G).\n",
        "Provide the output result fot the first 15 characters of the sequence.\n",
        "\n",
        "(3 Points)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "2V0qxMGfH5Cg"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Original (first 15 bases): GGTTAAACATGACTC\n",
            "Reverse complement: GAGTCATGTTTAACC\n"
          ]
        }
      ],
      "source": [
        "from Bio.Seq import Seq\n",
        "\n",
        "\n",
        "def reverse_complement_biopython(seq_str):\n",
        "\n",
        "    seq_str = seq_str.upper()\n",
        "    for base in seq_str:\n",
        "        if base not in \"ATCG\":\n",
        "            raise ValueError(\"Invalid nucleotide found. Only A, T, C, G are allowed.\")\n",
        "    \n",
        "    seq = Seq(seq_str)\n",
        "    return seq.reverse_complement()\n",
        "\n",
        "\n",
        "first_15 = str(record.seq[:15])\n",
        "rev_comp = reverse_complement_biopython(first_15)\n",
        "\n",
        "print(\"Original (first 15 bases):\", first_15)\n",
        "print(\"Reverse complement:\", rev_comp)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fzgsHpprH5KY"
      },
      "source": [
        "9. Write a function that takes your DNA sequence (first 21 characters) as input and performs both transcription and translation on a given DNA sequence. The function should:\n",
        "- Transcribe the DNA sequence into mRNA. In mRNA, thymine (T) is replaced by uracil (U).\n",
        "- Translate the mRNA sequence into a protein sequence using the standard genetic code. Do not forget about stop codons (they stop the translation).\n",
        "\n",
        "(4 Points)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "6dEBXJduH7Ij"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Original DNA (first 21 bases): GGTTAAACATGACTCCTAGAG\n",
            "Transcribed mRNA: GGUUAAACAUGACUCCUAGAG\n",
            "Translated Protein: G\n"
          ]
        }
      ],
      "source": [
        "from Bio.Seq import Seq\n",
        "\n",
        "\n",
        "def transcribe_and_translate(dna_seq):\n",
        "    \n",
        "    dna_seq = dna_seq.upper()\n",
        "    for base in dna_seq:\n",
        "        if base not in \"ATCG\":\n",
        "            raise ValueError(\"Invalid nucleotide found. Only A, T, C, G are allowed.\")\n",
        "    \n",
        "    # Create a Seq object\n",
        "    seq_obj = Seq(dna_seq)\n",
        "    \n",
        "    # Transcription: DNA → mRNA (replace T with U)\n",
        "    mrna_seq = seq_obj.transcribe()\n",
        "    \n",
        "    # Translation: mRNA → Protein (stop codons handled automatically)\n",
        "    protein_seq = mrna_seq.translate(to_stop=True)\n",
        "    \n",
        "    return mrna_seq, protein_seq\n",
        "\n",
        "\n",
        "\n",
        "first_21 = str(record.seq[:21])\n",
        "\n",
        "\n",
        "mrna, protein = transcribe_and_translate(first_21)\n",
        "\n",
        "\n",
        "print(\"Original DNA (first 21 bases):\", first_21)\n",
        "print(\"Transcribed mRNA:\", mrna)\n",
        "print(\"Translated Protein:\", protein)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qsyTtjWZH7PC"
      },
      "source": [
        "10. Write a function that calculates the total number of single-nucleotide substitutions (SNPs) in all possible codons (except STOP codons) that can result in a stop codon (TAA, TAG, or TGA).  Use Standard RNA codon table from [wikipedia](https://en.wikipedia.org/wiki/DNA_and_RNA_codon_tables).  \n",
        "Hints:  The simplest way to solve this task is to iterate over all possible codons and check how many single nucleotide changes (mutations) transform a sense codon into a stop codon.\n",
        "\n",
        "(4 Points)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "A1YZuIz9H9Em"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Total single-nucleotide substitutions (sense → stop): 23\n",
            "Number of sense codons with at least one nonsense SNP: 18\n",
            "\n",
            "Examples (sense codon → stop(s)):\n",
            "AAA -> UAA\n",
            "AAG -> UAG\n",
            "AGA -> UGA\n",
            "CAA -> UAA\n",
            "CAG -> UAG\n",
            "CGA -> UGA\n",
            "GAA -> UAA\n",
            "GAG -> UAG\n",
            "GGA -> UGA\n",
            "UAC -> UAA, UAG\n",
            "UAU -> UAA, UAG\n",
            "UCA -> UAA, UGA\n",
            "UCG -> UAG\n",
            "UGC -> UGA\n",
            "UGG -> UAG, UGA\n",
            "UGU -> UGA\n",
            "UUA -> UAA, UGA\n",
            "UUG -> UAG\n"
          ]
        }
      ],
      "source": [
        "\n",
        "\n",
        "from itertools import product\n",
        "\n",
        "def count_snp_to_stop_rna():\n",
        "    nts = \"AUCG\"  # RNA alphabet\n",
        "    stops = {\"UAA\", \"UAG\", \"UGA\"}\n",
        "\n",
        "    # Generate all codons\n",
        "    codons = [''.join(c) for c in product(nts, repeat=3)]\n",
        "    sense_codons = [c for c in codons if c not in stops]\n",
        "\n",
        "    total_snp_to_stop = 0\n",
        "    details = {}  # codon -> list of stop codons reachable by one SNP\n",
        "\n",
        "    def neighbors(c):\n",
        "        for i, base in enumerate(c):\n",
        "            for n in nts:\n",
        "                if n != base:\n",
        "                    yield c[:i] + n + c[i+1:]\n",
        "\n",
        "    for codon in sense_codons:\n",
        "        targets = sorted({n for n in neighbors(codon) if n in stops})\n",
        "        if targets:\n",
        "            details[codon] = targets\n",
        "            total_snp_to_stop += len(targets)\n",
        "\n",
        "    return total_snp_to_stop, details\n",
        "\n",
        "total, mapping = count_snp_to_stop_rna()\n",
        "\n",
        "print(\"Total single-nucleotide substitutions (sense → stop):\", total)\n",
        "print(\"Number of sense codons with at least one nonsense SNP:\", len(mapping))\n",
        "print(\"\\nExamples (sense codon → stop(s)):\")\n",
        "for codon, stops in sorted(mapping.items()):\n",
        "    print(f\"{codon} -> {', '.join(stops)}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VSEWm_nIB6yx"
      },
      "source": [
        "11. Please tell us an estimate on how long it took to complete this exercise (in hours, e.g. ~ 1.5 hours) and feel free to add any feedback (No points, optional)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oWbXbrkRB6yx"
      },
      "source": [
        "Estimated time to complete:\n",
        "It took approximately 1 hour to complete this exercise.\n",
        "\n",
        "Feedback:\n",
        "The theoretical concepts were clear and well-structured, making it easier to understand the underlying biological principles and apply them through coding. Overall, one hour was sufficient to review the theory and successfully implement all the required tasks."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": ".venv",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.12.2"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
